### DISCUSSION TITLE
user signup not possible?

### DISCUSSION THREAD

**Discussion by eikaramba** [directus/directus#10445] [Archive]
in the documentation there is no info about how to signup a user. login is possible, but signup not?

i found out that it might be possible by giving the public role the permission to create users and then make a POST request to /users. However the journey continues..

1) i don't get the created user in the result back. if i set read access to all then sure it works, but that is not intended as then every users profile would be public.
2) i set a field preset for the users create permission like this:
 ```
{
    "role": "c996783a-ff49-4686-84a3-d03c01a99940"
}
```

The idea is that each created user in the public role automatically gets the "user" role. However then i get a 403 upon trying to create the user.
hooks/index.js (also tried hooks/users/index.js)
```js
export default ({ filter, action }) => {
    console.log("users endpoint"); //never called
	filter('users.create', (input) => {
		console.log('Creating user', input); //never called
                // input.role = 'c996783a-ff49-4686-84a3-d03c01a99940';
	});
};
```

I think it would be really great to document the possible steps here and also how to set defaults. it seems the documentation is e.g. lacking how this whole field presets works, especially with relations.

And to make it even more complicated, i wasn't able to setup a custom filter hook for the users endpoint. otherwise i would have maybe set the role there before creating the user, but as i said, it's not called. i guess because the users endpoint is not living under /items and thus cannot have hooks?

**Comment by gokaybiz** [comment]
May you should check #3083

Edit:
The document link in #3083 seems dead, but if you understood the idea, you can make whatever you want by following these links; 

https://github.com/directus/directus/blob/25375cc4810b1e501cce1590ad8c21148ea596cb/api/src/services/users.ts#L134
https://github.com/directus/directus/blob/25375cc4810b1e501cce1590ad8c21148ea596cb/api/src/services/items.ts#L233

https://docs.directus.io/extensions/endpoints/#example-recipes

**Comment by eikaramba** [comment]
i saw that issue, that started my journey.

i was now able to solve at least the problem to create a hook for the user service. I needed to understand that the extensions needs to be commonjs modules and the documentation only gives es6 modules examples all complete with build scripts etc. makes sense but if you just quickly want to add some hooks you can just put the file into the folder, reload the server and it works.

hooks/users/index.js
```
module.exports = function registerHooks({ filter, action }) {
	filter('users.create', (input) => {
		input.role = 'c996783a-ff49-4686-84a3-d03c01a99940'; //replace with your own role
	});
};
```

I would argue this bit of information together with how to create users should be put into the documentation where also login etc. is explained. all other backend frameworks are doing it as well and although i like the approach directus is taking with the general users endpoint, one should then explain to use this for creating the user likewise to every other entity.

i can make a PR for that if you wish. otherwise feel free to close this issue or leave it open as long as this is not done

**Comment by rijkvanzanten** [comment]
There's currently no native "public registration" functionality through the regular login. You can already configure the SSO providers to create the user on first login.

See https://github.com/directus/directus/discussions/3083 for the discussion around adding native public registration to core 👍🏻

**Reply by trinzia** [reply]
A tutorial/video added to the Youtube tutorials would be great.  

As a front-end dev, I've read the documentation, watched the tutorials on Youtube, and am reading the Github issues/discussions, and I still don't know where to start, to create public registration and hopefully Gmail login.

**Reply by rijkvanzanten** [reply]
cc @erondpowell

**Reply by erondpowell** [reply]
> A tutorial/video added to the Youtube tutorials would be great.

Great suggestion, trinzia. It will take some time because there is a lot of documentation in queue right now, but this has been added.

**Reply by hgezim** [reply]
> You can already configure the SSO providers to create the user on first login

Any docs on this or more info on how I can do that?

**Reply by eikaramba** [reply]
@hgezim 
https://docs.directus.io/self-hosted/config-options.html#oauth-2-0
use the AUTH_<PROVIDER>_ALLOW_PUBLIC_REGISTRATION setting
